src/coreclr/jit/hwintrinsic.cpp

```diff
@@ -469,6 +469,215 @@ CorInfoType Compiler::getBaseJitTypeFromArgIfNeeded(NamedIntrinsic       intrins
+    {
+        // We should only have known SIMD sizes
+#if defined(TARGET_ARM64)
+        assert((info.simdSize == 8) || (info.simdSize == 16));
+#elif defined(TARGET_XARCH)
+        assert((info.simdSize == 16) || (info.simdSize == 32) || (info.simdSize == 64));
+#else
+        unreached();
+#endif
+    }
+
+    if (info.numArgs != -1)
+    {
+        // We should only have an expected number of arguments
+#if defined(TARGET_ARM64)
+        assert((info.numArgs >= 0) && (info.numArgs <= 4));
+#elif defined(TARGET_XARCH)
+        assert((info.numArgs >= 0) && (info.numArgs <= 5));
+#else
+        unreached();
+#endif
+    }
+
+    // TODO: There's more we could validate here in terms of flags, instructions used, etc.
+    // Some of this is already done ad-hoc elsewhere throughout the JIT
+}
+
+static void ValidateHWIntrinsicIsaRange(CORINFO_InstructionSet isa, const HWIntrinsicIsaRange& isaRange)
+{
+    // Both entries should be illegal if either is
+    if (isaRange.FirstId == NI_Illegal)
+    {
+        assert(isaRange.LastId == NI_Illegal);
+        return;
+    }
+    assert(isaRange.LastId != NI_Illegal);
+
+    // Both entries should belong to the expected ISA
+    assert(HWIntrinsicInfo::lookupIsa(isaRange.FirstId) == isa);
+    assert(HWIntrinsicInfo::lookupIsa(isaRange.LastId) == isa);
+
+    // The last ID should be the same as or after the first ID
+    assert(isaRange.FirstId <= isaRange.LastId);
+
+    // The ID before the range should not be part of the expected ISA
+    NamedIntrinsic prevId = static_cast<NamedIntrinsic>(isaRange.FirstId - 1);
+    assert((prevId == NI_HW_INTRINSIC_START) || (HWIntrinsicInfo::lookupIsa(prevId) != isa));
+
+    // The ID after the range should not be part of the expected ISA
+    NamedIntrinsic nextId = static_cast<NamedIntrinsic>(isaRange.LastId + 1);
+#if defined(TARGET_ARM64)
+    assert((nextId == NI_HW_INTRINSIC_END) || (HWIntrinsicInfo::lookupIsa(nextId) != isa) ||
+           (nextId == SPECIAL_NI_Sve));
+#else
+    assert((nextId == NI_HW_INTRINSIC_END) || (HWIntrinsicInfo::lookupIsa(nextId) != isa));
+#endif
+
+    NamedIntrinsic         ni       = static_cast<NamedIntrinsic>(isaRange.FirstId);
+    const HWIntrinsicInfo* prevInfo = &HWIntrinsicInfo::lookup(ni);
+    ValidateHWIntrinsicInfo(isa, ni, *prevInfo);
+
+    size_t count = (isaRange.LastId - isaRange.FirstId) + 1;
+
+    for (size_t i = 1; i < count; i++)
+    {
+        ni                          = static_cast<NamedIntrinsic>(isaRange.FirstId + i);
+        const HWIntrinsicInfo* info = &HWIntrinsicInfo::lookup(ni);
+        ValidateHWIntrinsicInfo(isa, ni, *info);
+
+        // The current name should be sorted after the previous
+        assert(strcmp(info->name, prevInfo->name) > 0);
+
+        prevInfo = info;
+    }
+}
+
+static void ValidateHWIntrinsicIsaRangeArray()
+{
+    size_t count = sizeof(hwintrinsicIsaRangeArray) / sizeof(hwintrinsicIsaRangeArray[0]);
+
+    for (size_t i = 0; i < count; i++)
```

Comment:
```suggestion
    for (size_t i = 0; i < ARRAY_SIZE(hwintrinsicIsaRangeArray); i++)
```

Reply:
👍, wasn't aware we had this helper. I'm used to `_countof` or similar but that hadn't been available for our Unix build.

Reply:
There is also `ArrLen` that we generally prefer to use in the JIT, but let's not rerun CI for that (looks like we have a small number of uses of `ARRAY_SIZE` anyway).

Reply:
Logged https://github.com/dotnet/runtime/issues/103908